### DISCUSSION TITLE
How to define UploadFile and form within a single pydantic model object to receive request data?

### DISCUSSION THREAD

**Discussion by TonyMistark** [fastapi/fastapi#11679] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from pydantic import BaseModel
from fastapi import UploadFile

class CostumUploadFile(UploadFile):
    def __init__(self, *args, check_filename: bool = False,**kwargs):
        self.check_filename = check_filename
    
    def validate(self, file: UploadFile):
        if self.check_filename:
            file.filename = self.file.filename + "from google"
        return file
# Solution 1
class MyModel(BaseModel):
    name: str
    file: CostumUploadFile(check_filename=True)


def check_filename_fun(file):
    file.filename = file.filename + "from google"
    return file
    

# Solution 2
class MyModel(BaseModel):
    name: str
    file: UploadFile = Field(check_filename=check_filename_fun)

@app.post("/upload")
async def upload_file(data: MyModel):
    # Access the uploaded file and other form data
    file = data.file
    name = data.name
    
    # Process the data as needed
    # ...
    
    return {"message": "File uploaded successfully"}
```

### Description

But the solution provided so far is:
```python

@app.post("/upload")
async def upload_file(file: UploadFile, name: Annotated[str, Form()]):
    
    # Process the data as needed
    # ...
    
    return {"message": "File uploaded successfully"}

```
This solution is still not convenient enough for me.

### Operating System

Linux

### Operating System Details

Distributor ID:	Ubuntu
Description:	Ubuntu 22.04.4 LTS
Release:	22.04
Codename:	jammy

### FastAPI Version

0.110.0

### Pydantic Version

2.7.2

### Python Version

Python 3.11.3

### Additional Context

In my application scenario, it is necessary to put the UploadFile and form data within a pydantic model. I need to perform validation by combining the UploadFile and form data, as this makes it more convenient to write the related business logic.

Additionally, I'm coming from Django REST Framework, where the serializer easily supports this kind of behavior. And for UploadFile, rewriting and adding custom validation for UploadFile is also more convenient and flexible. However, after switching to FastAPI, I'm unable to continue using these flexible features.

**Comment by YuriiMotov** [comment]
Pydantic models are not supported by FastAPI for Form parameters (and any other parameters, except Body) yet.
But looks like there are plans to implement it (FastAPI roadmap: https://github.com/tiangolo/fastapi/issues/10370).

As a workaround you can use Pydantic model, but you should specify the `__init__` method manually:
```

from typing import Annotated
from fastapi import Depends, FastAPI, Form, UploadFile


from pydantic import BaseModel, model_validator


class MyModel(BaseModel):
    name: Annotated[str, Form()]
    check_filename: Annotated[bool, Form()]
    file: UploadFile

    def __init__(
        self, name: Annotated[str, Form()], check_filename: Annotated[bool, Form()],
        file: UploadFile,
    ):
        super().__init__(name=name, check_filename=check_filename, file=file)

    @model_validator(mode="after")
    def validate(self):
        if self.check_filename:
            self.file.filename = self.file.filename + "from google"
        return self


app = FastAPI()


@app.post("/upload")
async def upload_file(data: MyModel = Depends()):
    # Access the uploaded file and other form data
    file = data.file
    name = data.name
    
    print(f"{file=}, {name=}")
    # Process the data as needed
   
    return {"message": "File uploaded successfully"}
```

**Reply by TonyMistark** [reply, marked as answer]
Thanks, it works now, although one field is written three times in `MyModel`, `__init__` and `super().__init__`, which is a bit redundant, but at least it is a runnable solution. Thanks again, I hope the official will soon provide a more convenient solution.

**Comment by TonyMistark** [comment]
[There](https://stackoverflow.com/a/65547551/24482100) and [There](https://stackoverflow.com/a/77113651/24482100) is a more elegant solution that is more suited to my use case, and I am currently using it.
```python
from fastapi.testclient import TestClient
from fastapi import FastAPI, Depends, Form, UploadFile
from pydantic import BaseModel


app = FastAPI()


def form_body(cls):
    new_params = [
        inspect.Parameter(
            field_name,
            inspect.Parameter.POSITIONAL_ONLY,
            default=model_field.default,
            annotation=Annotated[model_field.annotation, *model_field.metadata, Form()],
        )
        for field_name, model_field in cls.model_fields.items()
    ]

    cls.__signature__ = cls.__signature__.replace(parameters=new_params)

    return cls


@form_body
class Item(BaseModel):
    name: str
    another: str
    file: UploadFile


@app.post('/test', response_model=Item)
def endpoint(item: Item = Depends(Item)):
    return item


tc = TestClient(app)


r = tc.post('/test', data={'name': 'name', 'another': 'another'}, file)

assert r.status_code == 200
assert r.json() == {'name': 'name', 'another': 'another'}
```
